### PR TITLE
feat: Character level annotations for ui.text_annotator

### LINKED ISSUE #1357: Allow character level annotations in ui.text_annotator
#### Is your feature request related to a problem? Please describe

Currently, the ui.text_annotator widget has a smart algorithm included that extends the annotations to full words. This puts limitations on the granularity of the annotations.

If I have a text, that is badly formatted e.g. `This cityNYC<3 is awesome` and I only want the word `NYC` to be annotated, it's currently not possible. 

#### Describe the solution you'd like

Give an option to disable the automatic completion of words, and give annotators full control of what they are annotating. 

#### Describe alternatives you've considered

No alternatives

### PR DESCRIPTION
**Pull request by marek-mihok**
This PR introduces the ability to make free annotations without the smart selection. Simply click on the toggle to turn off the default smart selection.

https://user-images.githubusercontent.com/23740173/179707112-c0ad8223-e52a-4a48-ae57-32be4aace7cf.mov

Closes #1357

### CODE REVIEW

File: ui/src/text_annotator.tsx [hunk 1]
```diff
@@ -227,22 +265,53 @@ export
             <span style={removeIconStyle as React.CSSProperties} className={css.iconUnderlay}></span>
           </mark>
         )
-      }
+      },
+      handleOnMouseUp = (ev: React.MouseEvent<HTMLSpanElement>) => {
+        if (smart_selection) annotate()
+        else if (timeoutRef.current) annotate(ev.currentTarget) // double-click
+        else if (new Date().getTime() - mouseDownTimeRef.current > 250) annotate() // dragging (long click)
+        else timeoutRef.current = setTimeout(() => annotate(), 250) // click
```
Comment (mturoci) [inline ui/src/text_annotator.tsx:273]: Maybe this could be a better detection for double click? https://stackoverflow.com/questions/5497073/how-to-differentiate-single-click-event-and-double-click-event/53939059#53939059

File: ui/src/text_annotator.tsx [hunk 2]
```diff
@@ -130,31 +131,33 @@ export
...
   ),
-  XTextAnnotator = ({ model }: { model: TextAnnotator }) => {
+  XTextAnnotator = ({ model: { name, title, tags, items, trigger, readonly, smart_selection = true } }: { model: TextAnnotator }) => {
     const
-      [startIdx, setStartIdx] = React.useState<U | null>(null),
-      [activeTag, setActiveTag] = React.useState<S | undefined>(model.readonly ? undefined : model.tags[0]?.name),
+      [activeTag, setActiveTag] = React.useState<S | undefined>(readonly ? undefined : tags[0]?.name),
       [hoveredTagIdx, setHoveredTagIdx] = React.useState<U | null>(),
-      tagColorMap = model.tags.reduce((map, t) => {
+      tagColorMap = tags.reduce((map, t) => {
```
Comment (mturoci) [inline ui/src/text_annotator.tsx:138]: Let's use map instead of reduce: `tagColorMap = new Map(tags.map(t => [t.name, t.color]))` - more elegant.

File: ui/src/text_annotator.tsx [hunk 3]
```diff
@@ -173,26 +176,63 @@ export
...
         const tagToRemove = tokens[idx].tag
         for (let i = idx; tokens[i]?.tag === tagToRemove; i++) tokens[i].tag = undefined
         setTokens([...tokens])
         submitWaveArgs()
       },
-      updateStartIdx = (startIdx: U) => () => setStartIdx(startIdx),
-      annotate = (end: U) => () => {
-        const start = startIdx
-        if (start === null) return
+      timeoutRef = React.useRef<NodeJS.Timeout | null>(null),
```
Comment (mturoci) [inline ui/src/text_annotator.tsx:189]: The name is too generic.

File: ui/src/text_annotator.tsx [hunk 4]
```diff
@@ -173,26 +176,63 @@ export
...
+          selectedTextStr = selectedText.toString(),
+          startTokenIdx = parseInt(smart_selection ? startElData!.key! : startElData!.start!),
+          endTokenIdx = target ? (startTokenIdx + selectedTextStr.length) : parseInt(smart_selection ? endElData!.key! : endElData!.end!),
+          max = Math.max(startTokenIdx, endTokenIdx),
+          min = Math.min(startTokenIdx, endTokenIdx)
 
+        if (target && selectedTextStr.length) {
+          // smart selection on double click
+          const targetDataIdx = parseInt(target.dataset.key!)
+          let strLength = 0 // number of characters right to the selection target; if not 0, we loop to the left direction
```
Comment (mturoci) [inline ui/src/text_annotator.tsx:213]: The name is too generic, try to name it so that the first part of your comment is no longer necessary.

File: ui/src/text_annotator.tsx [hunk 5]
```diff
@@ -173,26 +176,63 @@ export
...
+          for (let i = 0; i <= selectedTextStr.length; i++) {
+            const tokenIdx = targetDataIdx + (!strLength ? i : (strLength - i))
```
Comment (mturoci) [inline ui/src/text_annotator.tsx:215]: Boolean negation makes conditionals hard to read, a lot of times simple if/else branch switch is enough.
```suggestion
            const tokenIdx = targetDataIdx + (strLength ? (strLength - i) : i)
```

File: ui/src/text_annotator.tsx [hunk 6]
```diff
@@ -173,26 +176,63 @@ export
...
+            tokens[tokenIdx].tag = activeTag
+            if (/[^a-z0-9]/i.test(tokens[tokenIdx + (!strLength ? 1 : -1)]?.text)) {
```
Comment (mturoci) [inline ui/src/text_annotator.tsx:217]:
```suggestion
            if (/[^a-z0-9]/i.test(tokens[tokenIdx + (strLength ? -1 : 1)]?.text)) {
```

**Review by mturoci — CHANGES_REQUESTED** [review]
Thanks @marek-mihok. Does this work reliably/consistently on Chrome/FF/Safari?

File: ui/src/text_annotator.tsx [hunk 7]
```diff
@@ -208,10 +243,8 @@ export
         return false
       },
       getMark = (text: S, idx: U, tag: S) => {
-        const color = tagColorMap.get(tag)
-        // Handle invalid tags entered by user
-        if (!color) return text
```
Comment (mturoci) [inline ui/src/text_annotator.tsx:213]: Why was this removed?

File: ui/src/text_annotator.tsx [hunk 8]
```diff
@@ -227,22 +260,47 @@ export
...
-          }
-        </div>
+      <div data-test={name} className={readonly ? css.readonly : ''}>
+        <div className={clas('wave-s16 wave-w6', css.title)}>{title}</div>
+        <AnnotatorTags tags={tags} activateTag={activateTag} activeTag={activeTag} />
+        <Fluent.Toggle
+          data-test={'toggle'}
+          label={'Smart selection'}
+          defaultChecked={smartSelection}
+          onChange={() => setSmartSelection(!smartSelection)}
```
Comment (mturoci) [inline ui/src/text_annotator.tsx:298]: Make sure to always get the most recent value.
```suggestion
          onChange={() => setSmartSelection(prevSelection => !prevSelection)}
```

File: ui/src/text_annotator.tsx [hunk 9]
```diff
@@ -227,22 +260,47 @@ export
...
-              <span key={idx} onMouseDown={updateStartIdx(idx)} onMouseUp={annotate(idx)}>{tag ? getMark(text, idx, tag) : text}</span>
-            ))
-          }
-        </div>
+      <div data-test={name} className={readonly ? css.readonly : ''}>
+        <div className={clas('wave-s16 wave-w6', css.title)}>{title}</div>
+        <AnnotatorTags tags={tags} activateTag={activateTag} activeTag={activeTag} />
+        <Fluent.Toggle
+          data-test={'toggle'}
+          label={'Smart selection'}
```
Comment (mturoci) [inline ui/src/text_annotator.tsx:296]: nit
```suggestion
          data-test='smart-selection'
          label='Smart selection'
```

File: ui/src/text_annotator.tsx [hunk 10]
```diff
@@ -227,22 +260,47 @@ export
...
+          key={idx}
+          idx={idx}
+          tokenProps={token}
+          handleMouseDown={handleMouseDown}
+          handleMouseUp={handleMouseUp}
+          handleMouseLeave={handleMouseLeave}
+          getMark={getMark}
+          activeColor={activeTag ? tagColorMap.get(activeTag)! : undefined}
+        />
+      }, [] as React.ReactElement[])
```
Comment (mturoci) [inline ui/src/text_annotator.tsx:284]: nit

```suggestion
      text = tokens.map((token, idx) => (<Token
          key={idx}
          idx={idx}
          tokenProps={token}
          handleMouseDown={handleMouseDown}
          handleMouseUp={handleMouseUp}
          handleMouseLeave={handleMouseLeave}
          getMark={getMark}
          activeColor={activeTag ? tagColorMap.get(activeTag) : undefined}
        />
      ))
```
Can also be part of the JSX - probably better.

File: ui/src/text_annotator.tsx [hunk 11]
```diff
@@ -130,31 +153,31 @@ export
...
-            arr.push({ text: textItem.substring(0, 1), tag })
-            arr.push({ text: textItem.substring(1), tag })
-          } else {
-            arr.push({ text: textItem, tag })
-          }
+      [smartSelection, setSmartSelection] = React.useState(true),
+      startElPropsRef = React.useRef<TokenMouseEventProps>(),
+      tagColorMap = new Map(tags.map(t => [t.name, t.color])),
+      [tokens, setTokens] = React.useState(items.reduce((arr, { text, tag }) => {
+        // If the smart_selection is True, split by any non-letter and non-number character.
```
Comment (mturoci) [inline ui/src/text_annotator.tsx:164]: Outdated comment.

**Review by mturoci — CHANGES_REQUESTED** [review]
Thanks @marek-mihok, looks much better!

text-annotator-character.png is probably leftover, right?

We can go over this tomorrow on a call.

File: ui/src/text_annotator.tsx (hunk 7 above)
Comment (marek-mihok) [reply ui/src/text_annotator.tsx:213]: Probably bad rebase since handling of invalid tags was implemented here: https://github.com/h2oai/wave/pull/1435

**Review by mturoci — APPROVED** [review]
